# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1437850503 14400
#      Sat Jul 25 14:55:03 2015 -0400
# Node ID 0fbe2e28ea84469a8f6632ea68a2a2001429d9fe
# Parent  39721c09691bc19cd045272fcaf94a30eacb8a05
build: Avoid triggering rebuild of docs when DOCSTRINGS regenerated (bug #45567)

* doc/module.mk: Make generated files depend on $(DOCSTRING_FILES) again.
(DOCSTRING_FILES) Delete, initialized elsewhere now.
(DOCSTRING_DEPS) Delete, no longer used.
* Makefile.am (DOCSTRING_FILES): Initialize to empty.
* libinterp/module.mk (DOCSTRING_FILES): Include libinterp/DOCSTRINGS in
the list.  (libinterp/DOCSTRINGS): New target with an order-only dependency
on libinterp/.DOCSTRINGS.
* scripts/module.mk (DOCSTRING_FILES): Include scripts/DOCSTRINGS in the
list.  (scripts/DOCSTRINGS): New target with an order-only dependency on
scripts/.DOCSTRINGS.

diff --git a/Makefile.am b/Makefile.am
--- a/Makefile.am
+++ b/Makefile.am
@@ -81,16 +81,17 @@ DOC_IMAGES_PNG =
 DOC_IMAGES_TXT =
 
 FCN_FILE_DIRS =
 FCN_FILES =
 GEN_FCN_FILES =
 PKG_ADD_FILES =
 SCRIPTS_IMAGES =
 JAR_FILES =
+DOCSTRING_FILES =
 
 bin_PROGRAMS =
 archlib_PROGRAMS =
 noinst_HEADERS =
 OCTAVE_VERSION_LINKS =
 OCTAVE_CROSS_TOOLS =
 OCTAVE_INTERPRETER_TARGETS =
 
diff --git a/doc/module.mk b/doc/module.mk
--- a/doc/module.mk
+++ b/doc/module.mk
@@ -298,26 +298,22 @@ doc-interpreter-dist-hook:
 	@$(GREP) '#define HAVE_CHOLMOD 1' $(top_builddir)/config.h > /dev/null || { echo "Documentation creation requires missing CHOLMOD library.  Cannot package distribution!" ; exit 1; }
 	@$(GREP) '#define HAVE_UMFPACK 1' $(top_builddir)/config.h > /dev/null || { echo "Documentation creation requires missing UMFPACK library.  Cannot package distribution!" ; exit 1; }
 	@$(GREP) '#define HAVE_QHULL 1' $(top_builddir)/config.h > /dev/null || { echo "Documentation creation requires missing QHULL library.  Cannot package distribution!" ; exit 1; }
 
 octetc_DATA += \
   doc/interpreter/doc-cache \
   doc/interpreter/macros.texi
 
-DOCSTRING_FILES = $(shell $(srcdir)/doc/interpreter/find-docstring-files.sh "$(top_srcdir)")
-
-DOCSTRING_DEPS = scripts/.DOCSTRINGS libinterp/.DOCSTRINGS
-
-doc/interpreter/doc-cache: $(DOCSTRING_DEPS) doc/interpreter/mk_doc_cache.m | $(OCTAVE_INTERPRETER_TARGETS) doc/interpreter/$(octave_dirstamp)
+doc/interpreter/doc-cache: $(DOCSTRING_FILES) doc/interpreter/mk_doc_cache.m | $(OCTAVE_INTERPRETER_TARGETS) doc/interpreter/$(octave_dirstamp)
 	$(AM_V_GEN)rm -f $@-t $@ && \
 	$(top_builddir)/run-octave -f -q -H $(srcdir)/doc/interpreter/mk_doc_cache.m - $(srcdir)/doc/interpreter/macros.texi $(DOCSTRING_FILES) > $@-t && \
 	mv $@-t $@
 
-$(MUNGED_TEXI_SRC): $(DOCSTRING_DEPS) $(munge_texi_SOURCES)
+$(MUNGED_TEXI_SRC): $(DOCSTRING_FILES) $(munge_texi_SOURCES)
 
 %.texi : %.txi doc/interpreter/munge-texi.pl | doc/interpreter/$(octave_dirstamp)
 	$(AM_V_GEN)rm -f $@-t $@ && \
 	$(PERL) $(srcdir)/doc/interpreter/munge-texi.pl $(top_srcdir) $(DOCSTRING_FILES) < $< > $@-t && \
 	mv $@-t $@
 
 doc/interpreter/contributors.texi: doc/interpreter/contributors.in | doc/interpreter/$(octave_dirstamp)
 	$(AM_V_GEN)rm -f $@-t $@ && \
diff --git a/libinterp/module.mk b/libinterp/module.mk
--- a/libinterp/module.mk
+++ b/libinterp/module.mk
@@ -290,16 +290,20 @@ DLDFCN_PKG_ADD_FILE = libinterp/dldfcn/P
 libinterp/dldfcn/PKG_ADD: $(DLDFCN_DEF_FILES) libinterp/mk-pkg-add
 	$(AM_V_GEN)rm -f $@-t && \
 	$(srcdir)/libinterp/mk-pkg-add $(DLDFCN_DEF_FILES) > $@-t && \
 	$(simple_move_if_change_rule)
 endif
 
 if AMCOND_BUILD_DOCS
 
+DOCSTRING_FILES += libinterp/DOCSTRINGS
+
+libinterp/DOCSTRINGS: | libinterp/.DOCSTRINGS
+
 libinterp/.DOCSTRINGS: $(ALL_DEF_FILES) libinterp/gendoc.pl
 	$(AM_V_GEN)rm -f $@-t $@ && \
 	if [ "x$(srcdir)" != "x." ] && [ -f $(srcdir)/libinterp/DOCSTRINGS ] && [ ! -f DOCSTRINGS ]; then \
 		cp $(srcdir)/libinterp/DOCSTRINGS libinterp/DOCSTRINGS; \
 		touch -r $(srcdir)/libinterp/DOCSTRINGS libinterp/DOCSTRINGS; \
 	fi && \
 	$(PERL) $(srcdir)/libinterp/gendoc.pl $(ALL_DEF_FILES) > $@-t && \
 	mv $@-t $@ && \
diff --git a/scripts/module.mk b/scripts/module.mk
--- a/scripts/module.mk
+++ b/scripts/module.mk
@@ -95,16 +95,20 @@ define GEN_FCN_FILES_TEMPLATE
 $(1): $(1:.m=.in) Makefile $(dir $(1))$(octave_dirstamp)
 	$$(AM_V_GEN)$$(do_subst_config_vals)
 endef
 
 $(foreach f, $(GEN_FCN_FILES), $(eval $(call GEN_FCN_FILES_TEMPLATE, $(f))))
 
 if AMCOND_BUILD_DOCS
 
+DOCSTRING_FILES += scripts/DOCSTRINGS
+
+scripts/DOCSTRINGS: | scripts/.DOCSTRINGS
+
 scripts/.DOCSTRINGS: $(FCN_FILES) $(GEN_FCN_FILES) scripts/mkdoc.pl Makefile
 	$(AM_V_GEN)rm -f $@-t $@ && \
 	if [ "x$(srcdir)" != "x." ] && [ -f $(srcdir)/scripts/DOCSTRINGS ] && [ ! -f scripts/DOCSTRINGS ]; then \
 		cp $(srcdir)/scripts/DOCSTRINGS scripts/DOCSTRINGS; \
 		touch -r $(srcdir)/scripts/DOCSTRINGS scripts/DOCSTRINGS; \
 	fi && \
 	$(PERL) $(srcdir)/scripts/mkdoc.pl "$(srcdir)" $(FCN_FILES) -- $(GEN_FCN_FILES) > $@-t && \
 	mv $@-t $@ && \
